### ISSUE TITLE
bundle lock --add-platform overrides the BUNDLED WITH

### ISSUE THREAD

**Issue by smsohan** [rubygems/rubygems#5683]
As per bundler [documentation](https://bundler.io/blog/2022/01/23/bundler-v2-3.html) 
> In Bundler 2.3 and up (if you also have RubyGems 3.3 or higher), running bundle install will use the exact version from the BUNDLED WITH section of the lockfile.

However, when bundle lock is used, instead of using the version of bundler as per Gemfile.lock, it uses the default bundler version and overrides the Gemfile.lock `BUNDLED WITH`. The documentation says the auto switch works for `bundle install`, but it'd be helpful to auto switch for the `lock` command as well. Here are the repro steps:

```bash
$ docker run -it --rm ruby bin/bash
root@2db1f7065bf7:/# mkdir /tmp/x
root@2db1f7065bf7:/# cd /tmp/x
root@2db1f7065bf7:/tmp/x# bundle -v
Bundler version 2.3.7

root@2db1f7065bf7:/tmp/x# bundle init
Writing new Gemfile to /tmp/x/Gemfile

root@2db1f7065bf7:/tmp/x# bundle
The Gemfile specifies no dependencies
Resolving dependencies...
Bundle complete! 0 Gemfile dependencies, 1 gem now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.

# system installed version of bundler is 2.3.7
root@2db1f7065bf7:/tmp/x# cat Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:

PLATFORMS
  x86_64-linux

DEPENDENCIES

BUNDLED WITH
   2.3.7

# imagine an app uses 2.1.4 instead
root@2db1f7065bf7:/tmp/x# sed -i "s/2.3.7/2.1.4/g" Gemfile.lock
root@2db1f7065bf7:/tmp/x# cat Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:

PLATFORMS
  x86_64-linux

DEPENDENCIES

BUNDLED WITH
   2.1.4

# We add a platform for this app, it should not override the version of bundler
root@2db1f7065bf7:/tmp/x# bundle lock --add-platform ruby
Resolving dependencies...
Writing lockfile to /tmp/x/Gemfile.lock

# But it overrides the Gemfile.lock version
root@2db1f7065bf7:/tmp/x# cat Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:

PLATFORMS
  ruby
  x86_64-linux

DEPENDENCIES

BUNDLED WITH
   2.3.7


# Expected output is as follows:

# But it overrides the Gemfile.lock version
root@2db1f7065bf7:/tmp/x# cat Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:

PLATFORMS
  ruby
  x86_64-linux

DEPENDENCIES

BUNDLED WITH
   2.1.4
```

This used to work before:

https://github.com/rubygems/bundler/blob/v2.1.4/lib/bundler/lockfile_generator.rb#L73-L75

But, now it uses the active bundler's version, which doesn't auto switch to the BUNDLED WITH version, like it does for install.

https://github.com/rubygems/rubygems/blob/ca63e3b21d3c5ca943d07f8584ccf29801b30218/bundler/lib/bundler/lockfile_generator.rb#L73-L75

**Comment by deivid-rodriguez** [comment]
Yes, what specific problem is this causing you?

In my opinion, the new behavior is more accurate because previously, even if Bundler 2.3.7 would have generated your `Gemfile.lock` file, the `BUNDLED WITH` section would still say `2.1.4`, which is not really true.

**Comment by deivid-rodriguez** [comment]
Closing due to lack of feedback.

**Comment by smsohan** [comment]
Here's the context. We build user code that gets submitted to Google Cloud serverless products. We want to add the Platform, but use the user specified bundler for the actual bundle install. Since this overrides the bundler version, we end up bundling with a different version of bundler than the one specified in the user's Gemfile.lock.

I see the reason this bug was fixed. But it seems like it could match the behaviour of bundle install and automatically choose the version of bundled based on what the user specifies.

**Comment by deivid-rodriguez** [comment]
I understand, and I'm not sure what's best.

On the one hand, `bundle lock --add-platform` is telling Bundler to add a platform to the lockfile, so I can see how it could be surprising that it changes the `BUNDLED WITH` section.

On the other hand, the `BUNDLED WITH` section is saying what version of Bundler last wrote the lockfile, so I can see how it could be surprising that bundling a lockfile with a certain version of bundler different from the one locked does not update it.

I don't think it would be a good idea to make `bundle lock` auto-install the `BUNDLED WITH` version and use that automatically, like `bundle install` does, because I think `bundle lock` actually installing something is more surprising than any of the things above.

So, in conclusion, I don't know 😅.
